### ISSUE TITLE
assert in spdk_bdev_io_get_buf related to alignment

### ISSUE THREAD

**Issue by jkryl** [spdk/spdk#713]
When creating AIO bdev following assert in spdk_bdev_io_get_buf (bdev.c) is triggered:

```
assert(len + alignment <= SPDK_BDEV_LARGE_BUF_MAX_SIZE + SPDK_BDEV_POOL_ALIGNMENT)
```

further analysis revealed that `spdk_bdev_get_buf_align()` returns bogus value (alignment). The alignment was 512k instead of 4k in my case. This function takes the alignment from `bdev->required_alignment`. So during the initialization of bdev something had to go wrong. And indeed, in `create_aio_disk()` at following line:

```
fdisk->disk.required_alignment = spdk_u32log2(block_size);
```
```
static inline uint32_t
spdk_u32log2(uint32_t x)
{
	if (x == 0) {
		/* log(0) is undefined */
		return 0;
	}
	return 31u - __builtin_clz(x);
}
```
The problem is `__builtin_clz` which in some cases (if cpu supports it) generates x86 assembler instruction LZCNT (count leading zeros in 32bit number). There is an interesting blog (https://nextmovesoftware.com/blog/2017/06/06/the-perils-of-using-__lzcnt-with-msvc/), which describes pitfalls of this instruction. Long story short if spdk is built on a cpu with ABM flag (advanced bit manipulation) and the resulting binary is deployed on a cpu without ABM flag, you will see the exact same mysterious failure as I did.

I understand that building and running spdk on different CPU is troublesome. Though DPDK does a very good job in this regard because it checks cpu flags when it starts and exits if not the same. SPDK lags behind DPDK in this regard. Even building a non-native DPDK with SPDK is a problem and cannot be done without modifying the makefiles .. Though I understand if this is something what's not going to be fixed.  The main purpose of this ticket is to make this issue known, so that people who might hit this problem in future, don't  spend their time finding the root cause.

## Context (Environment including OS version, SPDK version, etc.)

I was using spdk version 19.01. In master branch I have noticed that the assert has been removed (in 4b92ffb3f166fc92de293d780e76b7a7c13c3b89), which makes the situation probably even worse because the bug will go unnoticed and can cause more unexpected problems later in the IO path.

**Comment by changpe1** [comment]
[Bug Scrub]: @jkryl thanks for reporting, we should add support for different platforms, @benlwalker will take a look of this issue.

**Comment by benlwalker** [comment]
7dbdf87f6 allows for targeting specific x86 platforms when building, instead of always using "native". There are additional patches coming for full cross compilation support.

Once the compile-side is done, we'll still need a runtime check like DPDK has.

**Comment by jkryl** [comment]
@benlwalker thanks for the fix! 👍 I'm wondering how to pass this information to dpdk (if dpdk is built from the subrepository). There is this code where DPDK_CONFIG is set: https://github.com/spdk/spdk/blob/master/dpdkbuild/Makefile#L93 and it is set to native (always). If I explicitly set `TARGET_ARCHITECTURE` for spdk then dpdk will be still using native unless COMMON_CFLAGS propagate to dpdk and override DPDK_CONFIG somehow, which I'm not sure about.

Also I see some discrepancy between dpdk and spdk in a way how they set the march. DPDK has so called "machine profiles" which are profiles stored in `mk/machine/` (haswell, nehalem, silvermont, etc.) and based on selected "machine", gcc march (and mtune in some cases) gets set. spdk is unaware of these machine profiles and understands just TARGET_ARCHITECTURE (which is basically the march option). So is spdk going to align with dpdk on this?

**Comment by benlwalker** [comment]
For more context, see this whole series of patches for cross compilation: https://review.gerrithub.io/c/spdk/spdk/+/460977

But you're right - we need to pass `TARGET_ARCHITECTURE` down to DPDK somehow. Unfortunately, that's not trivial to do because we can't pass any `-march` value to DPDK. There's two major issues outstanding with our build system that need to be solved:
* We may need special flags/options for different architecture targets. Just setting `-march` isn't always enough, unfortunately. We already have a special case in there for ARM and POWER.
* We need to confirm that the system the executable is running on has all of the features of the target platform it was compiled for. That's a run time check that we just don't have anywhere today.

I don't have a concrete plan on how to do both of these just yet, although it's likely we'll follow DPDK's example. I really want to avoid duplicating all of the machine profile stuff from DPDK, but SPDK is very intentionally decoupled from DPDK via an abstraction layer and I'm not sure how to deal with that. I'll of course leave this bug open until we get it all fleshed out.

**Comment by jkryl** [comment]
Thanks for the information! Just to give a bit more info about our use case. Most of the people probably don't realise that this is not just a cross compilation problem (different cpu architectures) but even differences between more recent and less recent x86 cpus. So spdk which I compile on my laptop with intel corei7 processor cannot be run on my desktop with slightly older intel xeon processor.

In our fork of SPDK we have a temporary workaround. We introduced a new configure option which sets the DPDK machine profile. Of course it cannot be the final fix, but just to keep us going. The problem was somewhat pressing for us because we create spdk docker images and we want the docker image to work on all recent x86 cpus (let's say manufactured 10 years back). So for DPDK we used the new configure option `--with-dpdk-machine=nhm` and for SPDK we set following `CFLAGS=-march=corei7 -mno-movbe -mno-lzcnt -mno-bmi -mno-bmi2` to explicitly exclude some of the instructions which are problematic.

It's good that there is an ongoing effort in this area 👍
